### ISSUE TITLE
Incorrect payment status in few cases

### ISSUE THREAD

**Issue by knit-pay** [wp-pay-extensions/woocommerce#5]
Hello Team ( @remcotolsma and @rvdsteege )

I would like to bring it to your notice that for few cases wrong payment status gets updated. I will try to explain this step by step. let me know if you don't understand any steps.

1. customer selects Pronamic Pay payment method on the checkout page of woocommerce.
2. He tries to make the payment, the amount gets deducted from his account but the payment gets failed because it's not tracked in the payment gateway. (Let's say it Payment id 1). This makes the status of the order failed in WooCommerce.
3. Then the customer gets redirected to the fail payment screen where he sees the option to pay again.
4. Customer again selects Pronamic Pay payment method and attempts the payment again (Payment id 2). customer leaves the website without making the payment. This makes payment 2 status as open/pending and woocommerce order status as payment pending.
5. after some time pronamic_pay_payment_status_check runs for **payment id 1**, and payment gets successful because payment gateway got the success payment status. due to which woocommerce order status gets changed to "Processing".
6. After some time pronamic_pay_payment_status_check runs for **payment id 2**, and the payment gateway marked this payment as expired/failed/cancelled. Because of which Woocommerce order status gets marked as "failed" from "Processing". In actual Payment id 1 is already successfully, Payment id 2 should not change the status of this order to failed.

Proposed solution: Status of order should not get changed if the order has already reached final status ie Processing, Complete, Shipped status and some other payment tries to change the status. Only the payment which has made this order reached final status should be allowed to make it failed again if necessary.
What do you think about it? This is a very tricky and rare case, and I could not think of a better solution than this. What do you suggest?

**Comment by rvdsteege** [comment]
Hello Gautam! 👋

Could you please clarify the following:

> [...] the amount gets deducted from his account but the payment gets failed because it's not tracked in the payment gateway

Sounds like an issue with the payment gateway, as a successful payment should not result in a failed order?

Also, if there are multiple 'Pronamic payments' started for the same WooCommerce order, the order status will only be updated for status updates of the last payment:

https://github.com/wp-pay-extensions/woocommerce/blob/025f91f7fcd9d5e2e15ffb39123107e2d7f73345/src/Extension.php#L553-L558

**Comment by knit-pay** [comment]
Hello @rvdsteege 

I will try to explain the issue again in a little another way.

**Before I start explaining, let's consider these 2 things**
**(a)** This payment gateway marks the payment as fail/expire if payment does not get completed within provided time of let's say within 30 minutes.
**(b)** For some payment methods, payment status does not update to successful status instantly after successful payment. It takes time to update the payment status as successful after successful payment. Because sometimes response doesn't come instantly from the bank. Webhook is received after some time with payment status.

**Now I am explaining payment flow:**
1. Customer attempts to make payment (Payment ID 1) after adding product in cart and filling the details.
2. Customer fills in all the details and makes the payment and returns back to the WooCommerce website. While doing redirection, successful payment status **was not** received from the payment gateway. Thought this was a successful payment, but the success status was not received yet. The payment was still under pending status (or any other non-complete status). As I mentioned in point number (b) above, payment status update is not instant in few cases.
3. Customer sees WooCommerce page to attempt the payment again. The customer selects the payment gateway and attempts payment again (Payment ID 2). This time customer does **not** make a successful payment. The customer closes the tab/window and leaves the website.
4. After some time, the Website receives a webhook for Payment ID 1, with status payment successful. This makes the WooCommerce order status as Processing. Everything is fine till this step.
5. Issue is at this step. As mentioned in point number (a), this payment gateway marks the payment as failed/expire after some time if payment is **not** completed within some time. After some time pronamic_pay_payment_status_check or Webhook will mark payment ID 2 (last payment) as failed/expired because the customer has not completed this payment. Due to this, the WooCommerce Order status will be updated to Failed. In step number 4, the status of this order was already marked as Processing now this last payment should not mark it as failed again. Payment for this order is already made in Payment ID 1. Now, Payment ID 2 should not mark this order as failed.

In Short: the Below code allows the last payment to change the order status, whereas the last payment should not be allowed to change the order status if the order status was marked as processing by previous payments.
https://github.com/wp-pay-extensions/woocommerce/blob/025f91f7fcd9d5e2e15ffb39123107e2d7f73345/src/Extension.php#L553-L558

Proposed solution: Status of order should not get changed if the order has already reached final status ie Processing, Complete, Shipped status and some other payment tries to change the status. Only the payment which has made this order reached final status should be allowed to make it failed again if necessary.
